cleverhans/attacks.py

```diff
@@ -1,4 +1,416 @@
+            assert self.parse_params(params)
+            import tensorflow as tf
+            graph = tf.py_func(self.generate_np, [x], tf.float32)
+            self.inf_loop = False
+            return graph
+        else:
+            error = "No symbolic or numeric implementation of attack."
+            raise NotImplementedError(error)
+
+    def generate_np(self, x_val, params={}):
+        """
+        Generate adversarial examples and return them as a Numpy array. This
+        method should be overriden in any child class that implements an attack
+        that is not fully expressed symbolically.
+        :param x_val: A Numpy array with the original inputs.
+        :param params: Parameter dictionary used by child classes.
+        :return: A Numpy array holding the adversarial examples.
+        """
+        if self.back == 'th':
+            raise NotImplementedError('Theano version not implemented.')
+
+        if not self.inf_loop:
+            self.inf_loop = True
+            import tensorflow as tf
+
+            # Generate this attack's graph if not done previously
+            if not hasattr(self, "_x") and not hasattr(self, "_x_adv"):
+                input_shape = list(x_val.shape)
+                input_shape[0] = None
+                self._x = tf.placeholder(tf.float32, shape=input_shape)
+                self._x_adv = self.generate(self._x, params=params)
+            self.inf_loop = False
+        else:
+            error = "No symbolic or numeric implementation of attack."
+            raise NotImplementedError(error)
+
+        return self.sess.run(self._x_adv, feed_dict={self._x: x_val})
+
+    def parse_params(self, params):
+        """
+        Take in a dictionary of parameters and applies attack-specific checks
+        before saving them as attributes.
+        :param params: a dictionary of attack-specific parameters
+        :return: True when parsing was successful
+        """
+        return True
+
+
+class FastGradientMethod(Attack):
+    """
+    This attack was originally implemented by Goodfellow et al. (2015) with the
+    infinity norm (and is known as the "Fast Gradient Sign Method"). This
+    implementation extends the attack to other norms, and is therefore called
+    the Fast Gradient Method.
+    Paper link: https://arxiv.org/abs/1412.6572
+    """
+    def __init__(self, model, back='tf', sess=None, params={}):
+        """
+        Create a FastGradientMethod instance.
+        """
+        super(FastGradientMethod, self).__init__(model, back, sess, params)
+
+    def generate(self, x, params={}):
+        """
+        Generate symbolic graph for adversarial examples and return.
+        """
+        # Parse and save attack-specific parameters
+        assert self.parse_params(params)
+
+        if self.back == 'tf':
+            from .attacks_tf import fgm
+        else:
+            from .attacks_th import fgm
+
+        return fgm(x, self.model(x), y=self.y, eps=self.eps, ord=self.ord,
```

Comment:
We have a couple options for where to parse the parameters. 
* Initially we had them in the attack class init, which has the drawback of requiring different objects being instantiated for running the same attack multiple times. 
* Now we parse them when the user either calls `generate` or `generate_np`, which solves the problem above. I created the `parse_params` method so we can have the same consistent checks on the parameters regardless of which function is called first by the user.

Regarding the question of how the user is supposed to know what to pass for params. Right now, the user would probably first get an exception when he/she tries to run the attack without the right parameters. The exception is raised by `parse_params`, which gives the list of mandatory params and documents them in its docstring. However that may not be ideal and is actually related to the problem you raised in the next comment